### PR TITLE
Initial revision of GL_KHR_cooperative_matrix

### PR DESCRIPTION
**Pull request by jeffbolznv**
I've imported this as-is from the private repo, so we can do final review here.

### CODE REVIEW

File: extensions/khr/GLSL_KHR_cooperative_matrix.txt [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+
+Contact
+Neil Hickey, Arm (neil.hickey 'at' arm.com)
+
+Contributors
+Jeff Bolz, NVIDIA (jbolz 'at' nvidia.com)
+Kevin Petit, Arm (kevin.petit 'at' arm.com)
+
+Notice
+Copyright (c) 2022 The Khronos Group Inc. Copyright terms at http://www.khronos.org/registry/speccopyright.html
```
Comment (jeffbolznv) [inline extensions/khr/GLSL_KHR_cooperative_matrix.txt:12]:
```suggestion
Copyright (c) 2022-2023 The Khronos Group Inc. Copyright terms at http://www.khronos.org/registry/speccopyright.html
```

File: extensions/khr/GLSL_KHR_cooperative_matrix.txt [hunk 2]
```diff
@@ -0,0 +1,396 @@
+Name
+GL_KHR_cooperative_matrix
```
Comment (jeffbolznv) [inline extensions/khr/GLSL_KHR_cooperative_matrix.txt:2]: This file doesn't have the indentation that GLSL specs usually have. Should we fix?

File: extensions/khr/GLSL_KHR_cooperative_matrix.txt [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+uvec2/uvec4.
+
+
+Revision History
+
+Revision 1
+
+- Internal revisions.
+
+Revision 2
```
Comment (jeffbolznv) [inline extensions/khr/GLSL_KHR_cooperative_matrix.txt:392]: I suggest to delete rev2-3 and only document revisions that occur after publication.

File: extensions/khr/GLSL_KHR_cooperative_matrix.txt [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+of their element type and size. In the following functions, the generic type
+"coopmat" can accept an coopmat type with any type parameters. The
+"buf" arrays must be in either buffer storage or shared storage, and the
+array that is passed in can be sized or unsized.
+
+For all of these functions, for a given dynamic instance of the function
+call, all function parameters must be the same for all invocations in a
+given scope instance (where the scope is the scope the cooperative matrix
+type(s) were created with). All invocations in a given scope instance must
+be active or all must be inactive.
```
Comment (svenvh) [inline extensions/khr/GLSL_KHR_cooperative_matrix.txt:188]: This currently only applies to `coopMatLoad`, `coopMatStore` & `coopMatMulAdd`.  Do we need to say something similar in the "Expressions" section above?

File: extensions/khr/GLSL_KHR_cooperative_matrix.txt [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+For informational purposes (non-normative), the following is an
+expected way for an implementation to map GLSL constructs to SPIR-V
+constructs:
+
+coopmat -> OpTypeCooperativeMatrixKHR
+coopmat constructor from scalar value -> OpConstantComposite
+coopmat constructor from coopmat -> Op*Convert, OpConvert*To*
+coopmat.length() -> OpCooperativeMatrixLength
+coopmat[i] -> OpCompositeExtract/OpCompositeInsert/OpAccessChain
++, -, *, / -> OpFAdd, OpFNegate/OpFSub, OpMatrixTimesScalar, OpFDiv
```
Comment (kpet) [inline extensions/khr/GLSL_KHR_cooperative_matrix.txt:69]: We might want to list **OpFMul** as well. Also, is it worth adding the mapping for integer operations?

**Review by kpet — CHANGES_REQUESTED** [review]
A few minor things to address and this will IMO be good to go.

File: extensions/khr/GLSL_KHR_cooperative_matrix.txt (hunk 4 above)
Comment (jeffbolznv) [reply extensions/khr/GLSL_KHR_cooperative_matrix.txt:188]: A similar question came up at https://github.com/KhronosGroup/SPIRV-Registry/issues/208, see my response there. I'm not planning to make a change in this spec unless we decide something different there.

**Comment by jeffbolznv** [comment]
@kpet I think I've addressed all your feedback.

**Comment by nihui** [comment]
May I ask when this PR will be merged, I look forward to testing this new extension in neural network inference in the downstream vkpeak and ncnn projects, relying on the implementation of glslang and this change :smiley:

**Review by kpet — APPROVED** [review]
@jeffbolznv LGTM, thanks!

**Pull request merged**